### PR TITLE
Part1

### PR DESCRIPTION
**Pull request by AalexxJjamess**

### CODE REVIEW

File: Character.cpp [hunk 1]
```diff
@@ -98,20 +109,25 @@ void Character::attackInternal(Character& other)
...
 void Character::printStats()
 {
     std::cout << getName() << "'s stats: " << std::endl;
-    assert(false);
+    
     /*
     make your getStats() use a function from the Utility.h
     */
-    std::cout << getStats(); 
+    std::cout << getCharacterStats(this);;
```
Comment (matkatmusic) [inline Character.cpp:129]: make your derived class's `getStats()` function call a function from Utility.h
- [ ] revise

File: Character.cpp [hunk 2]
```diff
@@ -98,20 +109,25 @@ void Character::attackInternal(Character& other)
             b) your stats are boosted 10%
             c) the initial value of your stats is updated to reflect this boosted stat for the next time you defeat another character.
       */
-        assert(false);
+
+        restoreAndBuff(this->armor, *initialArmorLevel);
+        restoreAndBuff(this->hitPoints, *initialHitPoints);
+        restoreAndBuff(this->attackDamage, *initialAttackDamage);
```
Comment (matkatmusic) [inline Character.cpp:115]: you don't need `this->` here.
- [ ] revise

File: Character.h [hunk 3]
```diff
@@ -53,14 +53,11 @@ struct Character
...
 
-    void printStats()
-    {
-        std::cout << getName() << "'s stats: " << std::endl;
-        std::cout << getStats(); //make your getStats() use a function from the Utility.h
-        
-        std::cout << std::endl;
-        std::cout << std::endl;
-    }
+    void restoreAndBuff( int& stat, int& originalStat );
```
Comment (matkatmusic) [inline Character.h:56]: This function doesn't need to be a member function.  It is only called in the cpp file, so make it a free function in the cpp file.
- [ ] revise

File: DragonSlayer.cpp [hunk 4]
```diff
@@ -1,28 +1,47 @@
...
-        assert(false);
+        // assert(false);
         //DragonSlayers get a 10x boost when attacking dragons, from their attack item.
         //so they should USE their attack item before attacking the dragon... 
         //note that items are single-use only, so you need to reset it after use.  
         //look in the Character class for how the other item types are reset after use.
+
+        attackItem->use(this);
+        attackItem.reset();
+
```
Comment (matkatmusic) [inline DragonSlayer.cpp:33]: You should check if `attackItem` is holding a valid object first before using it.
- [ ] revise

File: Dwarf.cpp [hunk 5]
```diff
@@ -1,7 +1,24 @@
...
+std::string Dwarf::getStats()
+{
+    return getCharacterStats(this);
+}
+
+void Dwarf::attack(Character& other)
+{
+    //std::cout << name << " is attacking " << other.getName() << "!!!" << std::endl;
+    Character::attack(other);
+}
```
Comment (matkatmusic) [inline Dwarf.cpp:23]: This function doesn't add any functionality, so it should not be added.
- [ ] remove/revise

File: Paladin.cpp [hunk 6]
```diff
@@ -1,7 +1,22 @@
...
\ No newline at end of file
+std::string Paladin::getStats()
+{
+    return getCharacterStats(this);
+}
+
+void Paladin::attack(Character& other)
+{
+    Character::attack(other);
+}
```
Comment (matkatmusic) [inline Paladin.cpp:22]: This function doesn't add functionality to the base class implementation so it should not exist.
- [ ] revise.

File: Utility.cpp [hunk 7]
```diff
@@ -43,66 +42,66 @@ std::string getCharacterStats(Character* ch)
...
+    else if( auto* ch_1 = dynamic_cast<DragonSlayer*>(character))
     {
-        ch->boostArmor( item.getBoost() * 1.5 );
+        ch_1->boostArmor( static_cast<int>(item.getBoost() * 1.5 ));
     }
-    else if( auto* ch = dynamic_cast<Dragon*>(character) )
+    else if( auto* ch_2 = dynamic_cast<Dragon*>(character) )
     {
         //dragons don't need defensive items
     }
```
Comment (matkatmusic) [inline Utility.cpp:64]: lol help yourself with your variable names.
`auto* ds = dynamic_cast<DragonSlayer*>`
`auto* d = dynamic_cast<Dragon*>`..
etc
- [ ] revise

File: Utility.cpp [hunk 8]
```diff
@@ -43,66 +42,66 @@ std::string getCharacterStats(Character* ch)
...
+    else if( auto* ch2 = dynamic_cast<DragonSlayer*>(character))
     {
-        ch->boostHitPoints(item->getBoost() * 1.25);
+        ch2->boostHitPoints(static_cast<int>(item->getBoost() * 1.25));
     }
-    else if( auto* ch = dynamic_cast<Dragon*>(character) )
+    else if( auto* ch3 = dynamic_cast<Dragon*>(character) )
     {
         //dragons don't carry helpful items!
     }
```
Comment (matkatmusic) [inline Utility.cpp:83]: - [ ] revise these variable names so they help the reader understand what's going on.

File: Utility.cpp [hunk 9]
```diff
@@ -43,66 +42,66 @@ std::string getCharacterStats(Character* ch)
...
         //DragonSlayers get a 10x boost when attacking dragons, from their attack item.
         //so their attack item should boost their attack damage by a factor of 10
         //this means you need to GET the attack damage, multiply it by the item's boost, and BOOST the attackDamage with that multiplied value.  
         //check Character.h for available member functions you can use.
     }
-    else if( auto* ch = dynamic_cast<Dragon*>(character) )
+    else if( auto* _3ch = dynamic_cast<Dragon*>(character) )
     {
         //dragons don't carry attack items!
     }
```
Comment (matkatmusic) [inline Utility.cpp:106]: - [ ] revise (variable names)

File: Character.cpp (hunk 1 above)
Comment (AalexxJjamess) [reply Character.cpp:129]: Resolved. The functionality to call from Utility.h was already present in the derived classes' implementations - I just forgot to update to getStats in Character.cpp.

File: Utility.cpp (hunk 7 above)
Comment (AalexxJjamess) [reply Utility.cpp:64]: Done :p

File: DragonSlayer.cpp (hunk 4 above)
Comment (AalexxJjamess) [reply DragonSlayer.cpp:33]: Added an if statement to bypass the use of the item if the attackItem smart pointer is holding nullptr.

File: AttackItem.h [hunk 10]
```diff
@@ -8,4 +8,5 @@ struct AttackItem : Item
 {
     AttackItem() : Item("attackItem", 10) { }
     void use(Character* character) override;
+    bool hasBeenUsed;
```
Comment (matkatmusic) [inline AttackItem.h:11]: Remove this member variable, which is never used.
- [ ] revise

File: Dwarf.cpp [hunk 11]
```diff
@@ -16,9 +16,10 @@ std::string Dwarf::getStats()
     return getCharacterStats(this);
 }
 
-void Dwarf::attack(Character& other)
+/*void Dwarf::attack(Character& other)
 {
     //std::cout << name << " is attacking " << other.getName() << "!!!" << std::endl;
     Character::attack(other);
 }
+*/
```
Comment (matkatmusic) [inline Dwarf.cpp]: Remove dead code
- [ ] revise

File: Paladin.cpp [hunk 12]
```diff
@@ -16,7 +16,9 @@ std::string Paladin::getStats()
     return getCharacterStats(this);
 }
 
-void Paladin::attack(Character& other)
+/*void Paladin::attack(Character& other)
 {
     Character::attack(other);
 }
+*/
```
Comment (matkatmusic) [inline Paladin.cpp]: Remove dead code
- [ ] revise

File: AttackItem.h [hunk 13]
```diff
@@ -1 +1,11 @@
-#include "Item.h"
\ No newline at end of file
+#pragma once
+#include "Item.h"
+#include "Utility.h"
+
```
Comment (matkatmusic) [inline AttackItem.h:4]: You're not using anything from Utility in this header file.
Don't include what you don't use!
- [ ] revise

File: Character.cpp [hunk 14]
```diff
@@ -1,9 +1,11 @@
 #include "Character.h"
 #include <iostream>
 #include <vector>
-
+#include "assert.h"
 #include "DefensiveItem.h"
 #include "HelpfulItem.h"
+#include "Utility.h"
```
Comment (matkatmusic) [inline Character.cpp:7]: you're not using anything from `assert` in this file.
don't include what you don't use!
check the rest of the files you're including and make sure you need them.
- [ ] revise

File: DragonSlayer.cpp [hunk 15]
```diff
@@ -1,28 +1,54 @@
 #include "DragonSlayer.h"
 #include "Dragon.h"
+#include "Utility.h"
 
-//DragonSlayer::DragonSlayer
+#include "assert.h"
 
-//DragonSlayer::getName
+DragonSlayer::DragonSlayer(std::string name_, int hp, int armor_) :
```
Comment (matkatmusic) [inline DragonSlayer.cpp:7]: you're not using anything from `assert` in this file.
don't include what you don't use
- [ ] revise.

File: DragonSlayer.h [hunk 16]
```diff
@@ -1,8 +1,16 @@
 #pragma once
+#include "Character.h"
+#include "Utility.h"
+#include "AttackItem.h"
```
Comment (matkatmusic) [inline DragonSlayer.h:4]: you're not using anything from `Utility` in this file.
don't include what you don't use.
- [ ] revise

File: Dwarf.h [hunk 17]
```diff
@@ -1,7 +1,15 @@
 #pragma once
+#include "Character.h"
+#include "Utility.h"
```
Comment (matkatmusic) [inline Dwarf.h:4]: you're not using anything from Utility in this header file
don't include what you don't use
- [ ] revise

File: HelpfulItem.h [hunk 18]
```diff
@@ -10,6 +10,7 @@
 struct HelpfulItem : Item
```
Comment (matkatmusic) [inline HelpfulItem.h:10]: You don't need to include all of the things being included in this header file.
- [ ] revise

File: Paladin.h [hunk 19]
```diff
@@ -1,7 +1,15 @@
 #pragma once
+#include "Character.h"
+#include "Utility.h"
```
Comment (matkatmusic) [inline Paladin.h:3]: You don't need to include all of the things being included in this header file.
- [ ] revise

File: AttackItem.h (hunk 13 above)
Comment (AalexxJjamess) [reply AttackItem.h:4]: The definition and implementation of useAttackItem are included in Utility.h, the program breaks if I do not inlcude it here or in AttackItem.

Comment (AalexxJjamess) [reply AttackItem.h:4]: I have moved the include to the .cpp file

File: Dwarf.cpp [hunk 20]
```diff
@@ -15,3 +16,10 @@ std::string Dwarf::getStats()
 {
     return getCharacterStats(this);
 }
+
+void Dwarf::attack(Character& other)
+{
+    Character::attack(other);
+}
+
+
```
Comment (matkatmusic) [inline Dwarf.cpp]: This adds no additional functionality when `attack()` is called.  This does nothing other than add code bloat.
- [ ] remove it

File: Paladin.cpp [hunk 21]
```diff
@@ -15,3 +16,10 @@ std::string Paladin::getStats()
 {
     return getCharacterStats(this);
 }
+
+void Paladin::attack(Character& other)
+{
+    Character::attack(other);
+}
```
Comment (matkatmusic) [inline Paladin.cpp:23]: Again, this does not add any functionality to the base implementation when `attack()` is called.  Remove it
- [ ] revise

**Pull request merged**